api/node/index.ts

```diff
@@ -325,6 +341,243 @@ export class ArrayModel<T> extends Model<T> {
+*         return {
+*             first: name[1],
+*             last: name[0],
+*         };
+*     }
+* );
+*
+*
+* // prints "Emil, Hans"
+* console.log(mappedModel.rowData(0));
+*
+* // prints "Mustermann, Max"
+* console.log(mappedModel.rowData(1));
+*
+* // prints "Tisch, Roman"
+* console.log(mappedModel.rowData(2));
+*
+* // You can modifying the underlying {@link ArrayModel}:
+*
+* const model = new ArrayModel<Name>([
+*     {
+*         first: "Hans",
+*         last: "Emil",
+*     },
+*     {
+*         first: "Max",
+*         last: "Mustermann",
+*     },
+*     {
+*         first: "Roman",
+*         last: "Tisch",
+*     },
+* ]);
+*
+* const mappedModel = model.map(
+*     (data) => {
+*         return data.last + ", " + data.first;
+*     },
+*     (data) => {
+*         const name = data.split(", ");
+*         return {
+*             first: name[1],
+*             last: name[0],
+*         };
+*     }
+* );
+*
+* model.setRowData(1, { first: "Minnie", last: "Musterfrau" } );
+*
+* // prints "Emil, Hans"
+* console.log(mappedModel.rowData(0));
+*
+* // prints "Musterfrau, Minnie"
+* console.log(mappedModel.rowData(1));
+*
+* // prints "Tisch, Roman"
+* console.log(mappedModel.rowData(2));
+*
+* // Alternatively you can do it also on the MapModel
+*
+* const model = new ArrayModel<Name>([
+*     {
+*         first: "Hans",
+*         last: "Emil",
+*     },
+*     {
+*         first: "Max",
+*         last: "Mustermann",
+*     },
+*     {
+*         first: "Roman",
+*         last: "Tisch",
+*     },
+* ]);
+*
+* const mappedModel = model.map(
+*     (data) => {
+*         return data.last + ", " + data.first;
+*     },
+*     (data) => {
+*         const name = data.split(", ");
+*         return {
+*             first: name[1],
+*             last: name[0],
+*         };
+*     }
+* );
+*
+* mappedModel.setRowData(1, "Musterfrau, Minnie");
+*
+* // prints "Emil, Hans"
+* console.log(mappedModel.rowData(0));
+*
+* // prints "Musterfrau, Minnie"
+* console.log(mappedModel.rowData(1));
+*
+* // prints "Tisch, Roman"
+* console.log(mappedModel.rowData(2));
+ * ```
+ */
+export class MapModel<T, U> extends Model<U> {
+    #sourceModel: Model<T>;
+    #mapFunction: (data: T) => U;
+    #mapBackFunction: (data: U) => T;
+
+    /**
+     * Constructs the MapModel with a source model and map functions.
+     * @template T the type of the source model's items.
+     * @template U the type of the mapped items.
+     * @param sourceModel the wrapped model.
+     * @param mapFunction maps the data from T to U.
+     * @param mapBackFunction maps the data back from U to T.
+     */
+    constructor(
+        sourceModel: Model<T>,
+        mapFunction: (data: T) => U,
+        mapBackFunction: (data: U) => T
+    ) {
+        super();
+        this.#sourceModel = sourceModel;
+        this.#mapFunction = mapFunction;
+        this.#mapBackFunction = mapBackFunction;
+        this.notify = this.#sourceModel.notify;
```

Comment:
This makes me realize that our current interface of just one notify property is insufficient.

Suppose an array model is exposed to slint a slint repeater, and at the same time it is a source model for a map model, that in turn is used by another repeater.

If the array model changes, it calls self.notify.row_added for example. With the current design that will notify only either the first repeater or the map model, when in fact it must notify both.

In Rust and C++ the design allows for multiple "listeners" for model changes. We need to change our design in the Node.js port to follow the same pattern.

Instead of notify being a single hidden property, we should probably use a WeakSet of listeners.

@ogoffart what do you think?

Reply:
since, afaik, the `#notify` is mapping to a rust ModelNotify,  I think this should be fine since the rust's ModelNotify can be mapped to many models.
Unless I misunderstood the problem.

Reply:
Yes, that would be an option. But then every JS source model to a JS map model needs to be connected on the rust model notify side behind the scenes. Not sure if that’s worth it or of its easier to make the connection on the JS side.

Reply:
We will have the same thing with the filter model,  with the addition that there we'll have to do a mapping.

Reply:
Ok, let’s focus just on this PR (MapModel) and the problem I see with the line we’re commenting on.

Setup:

“sourceModel” is a JS ArrayModel
“mapModel” is a JS MapModel which uses “sourceModel” as source model when constructed.

Scenario 1:

“sourceModel” is created where “notify” is a NullPeer, “mapModel” is created, the notify property is copied, thus both share a reference to the same NullPeer. “mapModel” is assigned to a Slint model property, with say a ListView using it.

This allocates a new JsModel, ModelRc, and JsSlintModelNotify. The notify property on the “mapModel” is written to point to the JsSlintModelNotify instance.

Any change to the “sourceModel” is “reported” into to NullPeer.  The ListView is not notified of any changes.

Scenario 2:

“sourceModel” is created, set on a Slint model property. The NullPeer is replaced by a JsSlintModelNotify. “mapModel” is created, the notify reference is copied. “mapModel” is set on a Slint model property, and the “mapModel”’s “notify” property is replaced by a new JsSlintModelNotify. Any changes to “sourceModel” are reflected in say a ListView that used “sourceModel”, but not by a ListView using “mapModel”, because that has its own JsSlintModelNotify, which nobody calls.


These are two scenarios that I believe need to work and should be covered by tests.

I don’t feel strongly how this is implemented. One low-effort way specifically for this map model situation would be to emulate the Rust MapModel way: Provide a setter _and_ getter for the notify property that forwards to the source model, and avoid unconditional creation of JsSlintModelNotify instances but instead re-use existing instances.

Does this make sense? Is this actionable?

Reply:
I can otherwise offer a plan B: We take this PR in as-is (without the working notify) and I'll try to rework the notify mechanism, also to make use of the fact that we have a base class and therefore implement it like in C++.

Let me know what your preferences are.

Reply:
@tronical ok I think it make sense to fix the notfiy issue in a separate pr. I can also do the implementation.

Reply:
Alright, sounds good